### PR TITLE
Added the default empty value for optional selects

### PR DESCRIPTION
**Pull request by Seb-C**

### CODE REVIEW

**Review by jguyomard — APPROVED** [review]

**Review by jguyomard — CHANGES_REQUESTED** [review]
There is a problem: the "empty" value will shift the indexes. So the values displayed in the results will be false (only the key is saved in database).

**Review by jguyomard — APPROVED** [review]

File: classes/controller/admin/form.ctrl.php [hunk 1]
```diff
@@ -67,8 +67,8 @@ public function before_save($item, $data)
             throw new \Exception(__('Your form must have at least one field.'));
         }
 
-        // Formats fields data
-        foreach ($fields_data as $field_id => $field_data) {
+        // Validating selected default value for fields with multiple choices
+        foreach ($fields_data as $field_id => &$field_data) {
```
Comment (shaoshiva) [inline classes/controller/admin/form.ctrl.php:71]: `$field_id` is no longer necessary.

File: classes/controller/admin/form.ctrl.php [hunk 2]
```diff
@@ -67,8 +67,8 @@ public function before_save($item, $data)
             throw new \Exception(__('Your form must have at least one field.'));
         }
 
-        // Formats fields data
-        foreach ($fields_data as $field_id => $field_data) {
+        // Validating selected default value for fields with multiple choices
```
Comment (shaoshiva) [inline classes/controller/admin/form.ctrl.php:70]: The comment is misleading. You're not doing validation but formatting (or morphing or transforming or whatever you want that indicates the right behaviour).

File: classes/controller/admin/form.ctrl.php [hunk 3]
```diff
@@ -98,7 +99,7 @@ public function before_save($item, $data)
 
         // Registers the fields
         $fieldsConfig = \Arr::get($this->config, 'fields_config.fields');
-        foreach ($fields_data as $field_id => $field_data) {
+        foreach ($fields_data as $field_id => &$field_data) {
```
Comment (shaoshiva) [inline classes/controller/admin/form.ctrl.php:102]: Why this change ? It seems not necessary.

File: static/src/js/admin/field/select.js [hunk 4]
```diff
@@ -15,40 +15,51 @@ define(['jquery-nos'], function($) {
...
 
             // Creates the select
             var $new = $('<select />').attr({
                 name: $default_value.attr('name'),
                 id: $default_value.attr('id')
             });
 
+            if (!$mandatory.prop('checked')) {
+                $new.append(
+                    $('<option value="" ' + (default_value_value === "" ? 'selected' : '') + '></option>')
```
Comment (shaoshiva) [inline static/src/js/admin/field/select.js:39]: You should check if there isn't already an empty option before injecting one, otherwise you could have a double empty option.

File: classes/controller/admin/form.ctrl.php (hunk 2 above)
Comment (Seb-C) [reply classes/controller/admin/form.ctrl.php:70]: The purpose of this part of the code is only to validate that the default value from inputs is a valid and existing value from the available options. There is no other side effects.

File: classes/controller/admin/form.ctrl.php (hunk 3 above)
Comment (Seb-C) [reply classes/controller/admin/form.ctrl.php:102]: Because the same variable name is used in the two foreach. And since the first foreach loops on the reference, not doing the same on the second one causes a weird bug (PHP replaces the content referenced by the variable instead).

File: static/src/js/admin/field/select.js (hunk 4 above)
Comment (Seb-C) [reply static/src/js/admin/field/select.js:39]: I don't think this would be coherent. The added option represents the absence of value (`null`), while a manually added option is still actually an option (non-`null` value). And this script is necessary to allow the default value of the select to be `null` when the field is optional.

File: classes/controller/admin/form.ctrl.php (hunk 3 above)
Comment (shaoshiva) [reply classes/controller/admin/form.ctrl.php:102]: Wow that's really weird. Ok.

File: static/src/js/admin/field/select.js (hunk 4 above)
Comment (shaoshiva) [reply static/src/js/admin/field/select.js:39]: What would not be coherent is that after merging this PR, all the select fields with an empty option and the "Mandatory" field checked will have a double empty choice in the default selected values.

File: classes/controller/admin/form.ctrl.php (hunk 2 above)
Comment (shaoshiva) [reply classes/controller/admin/form.ctrl.php:70]: I do not agree.

validate = check or prove the validity or accuracy of.
format = (especially in computing) arrange or put into a format.
